src/server/service/kv.rs

```diff
@@ -859,31 +783,25 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
     ) {
         let timer = GRPC_MSG_HISTOGRAM_VEC.raw_delete_range.start_coarse_timer();
 
-        let (cb, future) = paired_future_callback();
+        let (cb, f) = paired_future_callback();
         let res = self.storage.async_raw_delete_range(
             req.take_context(),
             req.take_cf(),
             req.take_start_key(),
             req.take_end_key(),
             cb,
         );
-        if let Err(e) = res {
-            self.send_fail_status(ctx, sink, Error::from(e), RpcStatusCode::ResourceExhausted);
-            return;
-        }
 
-        let future = future
-            .map_err(Error::from)
-            .map(|v| {
+        let future = AndThenWith::new(res, f.map_err(Error::from))
```

Comment:
How about:

```rust
future::result(res)
    .map_err(Error::from)
    .and_then(|_| future.map_err(Error::from))
    .and_then(|...|)
```

like what we did in [storage](https://github.com/tikv/tikv/blob/master/src/storage/mod.rs#L547)?

Reply:
That's wrong. `res` should be handled the same way as what return by the `future`. So this becomes `future::result(res).and_then(|_| future.flatten())` like what is [commented](https://github.com/tikv/tikv/pull/3450/files#diff-ce14a2c92d99351d9bbb883addee9878R35). All in all, using a custom future here is more expressive and clean.

Reply:
It's approachable but I prefer this way as it's shorter.